### PR TITLE
test: always get a working AKS version

### PR DESCRIPTION
**Pull request by jackfrancis**
**What type of PR is this?**

/kind other

This is an evolution of the AKS E2E test foo.

**What this PR does / why we need it**:

This PR tweaks slightly the functional outcome of the `GetAKSKubernetesVersion` convenience function, use by the E2E runner to ensure that we always test AKS clusters with a working Kubernetes version in the following ways:

- Now tolerates the edge case where the passed in version is higher than an actual working AKS Kubernetes version, and that AKS Kubernetes version is a ".0" release (i.e., the initial release of a new minor version).

That function has also been decomposed to rely upon a new, exported `GetWorkingAKSKubernetesVersion` function, which has less capz E2E context and so can be re-used by other projects testing AKS in capz.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/aks.go [hunk 1]
```diff
@@ -225,46 +225,72 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
+	return []client.ListOption{
+		client.InNamespace(namespace),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: name,
+		},
+	}
+}
+
+// GetWorkingAKSKubernetesVersion returns an available Kubernetes version of AKS given a desired semver version, if possible
+// If the desired version is available, we return it;
```
Comment (CecileRobertMichon) [inline test/e2e/aks.go:245]: thoughts or incorporating something like https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1767/files#diff-35f348a5d567ebeba39b1e40bbab0e1a7f64009cbadff142952f9b9691e94666R137 for "give me the latest version of this Major.Minor release" ? The previous behavior was to get the latest patch for any valid version but now we're just going to stick to that version if it is valid so I think we should also preserve the existing behavior of "latest available", ideally using the same input format as the other tests, ie. "stable-major.minor".

File: test/e2e/config/azure-dev.yaml [hunk 2]
```diff
@@ -97,7 +97,7 @@ providers:
 
 variables:
   KUBERNETES_VERSION: "${KUBERNETES_VERSION:-v1.22.1}"
-  AKS_KUBERNETES_VERSION: "${KUBERNETES_VERSION:-v1.19.9}"
+  AKS_KUBERNETES_VERSION: "${KUBERNETES_VERSION:-stable-1.19}"
```
Comment (CecileRobertMichon) [inline test/e2e/config/azure-dev.yaml:100]: ❤️

File: test/e2e/aks.go [hunk 3]
```diff
@@ -225,46 +225,81 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
 	Expect(err).NotTo(HaveOccurred())
 	subscriptionID := settings.GetSubscriptionID()
-	authorizer, err := settings.GetAuthorizer()
+	maxVersion, err := GetWorkingAKSKubernetesVersion(ctx, subscriptionID, location, e2eAKSVersion)
 	Expect(err).NotTo(HaveOccurred())
+
+	return maxVersion, nil
+}
+
+// byClusterOptions returns a set of ListOptions that allows to identify all the objects belonging to a Cluster.
```
Comment (mboersma) [inline test/e2e/aks.go:234]:
```suggestion
// byClusterOptions returns a set of ListOptions that will identify all the objects belonging to a Cluster.
```

File: test/e2e/aks.go [hunk 4]
```diff
@@ -225,46 +225,81 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
+func byClusterOptions(name, namespace string) []client.ListOption {
+	return []client.ListOption{
+		client.InNamespace(namespace),
+		client.MatchingLabels{
+			clusterv1.ClusterLabelName: name,
+		},
+	}
+}
+
+// GetWorkingAKSKubernetesVersion returns an available Kubernetes version of AKS given a desired semver version, if possible
```
Comment (mboersma) [inline test/e2e/aks.go:244]:
```suggestion
// GetWorkingAKSKubernetesVersion returns an available Kubernetes version of AKS given a desired semver version, if possible.
```

File: test/e2e/aks.go [hunk 5]
```diff
@@ -225,46 +225,81 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
+// If the desired version is available, we return it;
```
Comment (mboersma) [inline test/e2e/aks.go:245]:
```suggestion
// If the desired version is available, we return it.
```

File: test/e2e/aks.go [hunk 6]
```diff
@@ -225,46 +225,81 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
+// If the desired version is not available, we check for any available patch version using desired version's Major.Minor semver release
```
Comment (mboersma) [inline test/e2e/aks.go:246]:
```suggestion
// If the desired version is not available, we check for any available patch version using desired version's Major.Minor semver release.
```

File: test/e2e/aks.go [hunk 7]
```diff
@@ -225,46 +225,81 @@ func GetAKSKubernetesVersion(ctx context.Context, e2eConfig *clusterctl.E2EConfi
...
+// If no versions are available in the desired version's Major.Minor semver release, we return an error
```
Comment (mboersma) [inline test/e2e/aks.go:247]:
```suggestion
// If no versions are available in the desired version's Major.Minor semver release, we return an error.
```

**Review by mboersma — APPROVED** [review]
/lgtm

Code looks good. I just had some minor punctuation suggestions.

**Review by CecileRobertMichon — COMMENTED** [review]
lgtm

**Comment by CecileRobertMichon** [comment]
@jackfrancis please address the comments above when you have a chance so we can merge this

**Review by mboersma — APPROVED** [review]
/lgtm
